### PR TITLE
feat(client): Implement basic list API

### PR DESCRIPTION
**Pull request by bergundy**

### CODE REVIEW

File: packages/client/src/workflow-client.ts [hunk 1]
```diff
@@ -946,6 +924,31 @@ export class WorkflowClient {
...
+  public async *list(query: string, options?: ListOptions): AsyncIterable<WorkflowExecutionInfo> {
+    let nextPageToken: Uint8Array = Buffer.alloc(0);
+    for (;;) {
+      const respose = await this.workflowService.listWorkflowExecutions({
+        namespace: this.options.namespace,
+        query,
+        nextPageToken,
+        pageSize: options?.pageSize,
+      });
+      // Not decoding memo payloads concurrently even though we could have to keep the lazy nature of this iterator.
```
Comment (Sushisource) [inline packages/client/src/workflow-client.ts:944]: Any particular reason?

**Review by Sushisource — APPROVED** [review]
Cool. Straightforward, works for me.

File: packages/client/src/workflow-client.ts (hunk 1 above)
Comment (bergundy) [reply packages/client/src/workflow-client.ts:944]: It's not worth it IMHO for now, memos tend to generally be small.

File: packages/client/src/workflow-client.ts [hunk 2]
```diff
@@ -946,6 +926,31 @@ export class WorkflowClient {
...
+      for (const raw of respose.executions) {
+        yield await executionInfoFromRaw(raw, this.dataConverter);
+      }
+      nextPageToken = respose.nextPageToken;
```
Comment (lorensr) [inline packages/client/src/workflow-client.ts:950]:
```suggestion
      const response = await this.workflowService.listWorkflowExecutions({
        namespace: this.options.namespace,
        query,
        nextPageToken,
        pageSize: options?.pageSize,
      });
      // Not decoding memo payloads concurrently even though we could have to keep the lazy nature of this iterator.
      for (const raw of response.executions) {
        yield await executionInfoFromRaw(raw, this.dataConverter);
      }
      nextPageToken = response.nextPageToken;
```

**Review by lorensr — APPROVED** [review]

File: packages/client/src/workflow-client.ts [hunk 3]
```diff
@@ -946,6 +926,31 @@ export class WorkflowClient {
...
   }
+
+  /**
+   * List workflows by given `query`.
+   *
+   * See more info on how to build the query in the Temporal docs: https://docs.temporal.io/visibility
+   *
+   * This API requires Elasticsearch based visibility as of server version 1.18.
+   */
+  public async *list(query: string, options?: ListOptions): AsyncIterable<WorkflowExecutionInfo> {
```
Comment (cretz) [inline packages/client/src/workflow-client.ts:937]: Should query be a list option or at least optional?

File: packages/client/src/workflow-client.ts (hunk 2 above)
Comment (bergundy) [reply packages/client/src/workflow-client.ts:950]: Me and my typos :(

File: packages/client/src/workflow-client.ts [hunk 4]
```diff
@@ -946,6 +926,31 @@ export class WorkflowClient {
       followRuns: options?.followRuns ?? true,
     });
   }
+
+  /**
+   * List workflows by given `query`.
+   *
+   * See more info on how to build the query in the Temporal docs: https://docs.temporal.io/visibility
+   *
+   * This API requires Elasticsearch based visibility as of server version 1.18.
```
Comment (cretz) [inline packages/client/src/workflow-client.ts:935]: Not exactly true as simple queries, e.g. listing open workflows, work w/out elastic search

File: packages/client/src/workflow-client.ts (hunk 3 above)
Comment (cretz) [inline packages/client/src/workflow-client.ts:937]: If describe is interceptable, this should be right?

File: packages/client/src/workflow-client.ts (hunk 4 above)
Comment (bergundy) [reply packages/client/src/workflow-client.ts:935]: I didn't know that, I thought this entire API only works with ES right now.
I'll fix.

File: packages/client/src/workflow-client.ts (hunk 3 above)
Comment (bergundy) [reply packages/client/src/workflow-client.ts:937]: Good question... is it optional on the server?

Comment (bergundy) [reply packages/client/src/workflow-client.ts:937]: I don't see any point making either of them interceptable.

Comment (lorensr) [reply packages/client/src/workflow-client.ts:937]: think so

image

Comment (cretz) [reply packages/client/src/workflow-client.ts:937]: I support either approach, but the one-is-one-isn't seems a little strange. Today describe is interceptable.

Comment (bergundy) [reply packages/client/src/workflow-client.ts:937]: Looks like it's optional on the server..

File: packages/client/src/workflow-client.ts (hunk 4 above)
Comment (bergundy) [reply packages/client/src/workflow-client.ts]: Thanks, fixed.

File: packages/client/src/workflow-client.ts (hunk 3 above)
Comment (bergundy) [reply packages/client/src/workflow-client.ts]: Made `query` optional.

Comment (bergundy) [reply packages/client/src/workflow-client.ts]: Tracking this in a separate issue.
I want to clean up client interceptors a bit before adding this.

**Pull request merged**